[FLINK-7068][blob] change BlobService sub-classes for permanent and transient BLOBs

NicoK opened the pull request:
This splits up the BLOB store classes into two different use cases which is the biggest change of FLIP-19:
* a `PermanentBlobStore` should resemble use cases for BLOBs that are permanently stored for a job's life time (HA and non-HA).
* a `TransientBlobStore` should reflect BLOB offloading for logs, RPC, etc. which even does not have to be reflected by files.

The `BlobServer` implements both while the `BlobCache` provides access to individual implementations of the two. HA files may only be uploaded by the `BlobServer` (and the (now) only publicly useful method of `BlobClient` to upload jar files). All other accesses should go through the two BLOB store classes.

This PR is based upon #4238 in a series to implement FLINK-6916. It contains some development history and should be squashed.

- [X] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [X] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [X] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -0,0 +1,391 @@
+					final File localFile =
+						new File(BlobUtils.getStorageLocationPath(storageDir.getAbsolutePath(), jobId));
+					try {
+						FileUtils.deleteDirectory(localFile);
```
note to self: this should also be guarded by the `readWriteLock.writeLock()`

tillrohrmann commented:
Hi Nico, I'm still waiting for Travis before merging FLINK-7056. In the meantime could you rebase this PR and squash all FLINK-7068 commits into one? Thanks.

tillrohrmann commented:
I've merged #4238. Could you please rebase onto the latest master? Moreover, I've noticed that `JobManagerCleanupITCase.testBlobServerCleanupCancelledJob` failed in one of the Travis runs. Not sure if this is related.

NicoK commented:
Rebased onto `master`, but had to drag in #4402 early to fix the end-to-end tests failing due to spurious warnings. The test failure you observed was actually a test instability introduced with #4238 for which I added a hotfix to this PR.

NicoK commented:
sorry for the mess, but let me also drag in #4568 and adapt the code in here (which is moved from `BlobCache` to `PermanentBlobCache` by this PR)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -0,0 +1,429 @@
+ * If files for a job are not needed any more, they will enter a staged, i.e. deferred, cleanup.
+ * Files may thus still be be accessible upon recovery and do not need to be re-downloaded.
+ */
+public class PermanentBlobCache extends TimerTask implements PermanentBlobService {
```
Just a wild thought: I noticed that the `TransientBlobCache` and the `PermanentBlobCache` have a lot of code in common. In order to reduce code duplication couldn't we create a common base class or let `PermanentBlobCache` extend `TransientBlobCache` adding the ref counting?

NicoK replied:
The latter is not possible since each class has its uniquely offered methods, e.g. `TransientBlobCache` allows deleting BLOBs, while `PermanentBlobCache` does not. I can come up with a common base class though.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -174,7 +176,6 @@ public ServerSocket createSocket(int port) throws IOException {
 		// start the server thread
 		setName("BLOB Server listener at " + getPort());
 		setDaemon(true);
-		start();
```
Why did you pull `start` out of the constructor? Wouldn't one always want to start the `BlobServer` when creating it?

NicoK replied:
Basically yes but this pattern does not work well with classes inheriting from `BlobServer`. Here, this is only the case for the `TestingFailingBlobServer` which sets additional properties and relies on them being set while executing `run()`. If started in the constructor, the `run()` method would work on a partially initialised object.

tillrohrmann replied:
Makes sense.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobService.java:
```diff
@@ -18,61 +18,31 @@
+	 * @return BLOB service
 	 */
-	File getFile(JobID jobId, BlobKey key) throws IOException;
+	PermanentBlobService getPermanentBlobStore();
```
I'm not so sure about the naming here. The `BlobStore` is actually something else than the `BlobService`. Would be good if we use a consistent naming for the different things (this could also include renaming some of the entities).

NicoK commented on flink-fs-tests/src/test/java/org/apache/flink/hdfstests/HDFSTest.java:
```diff
@@ -257,7 +265,88 @@ public void testBlobServerRecovery() throws Exception {
+		org.apache.flink.configuration.Configuration
+			config = new org.apache.flink.configuration.Configuration();
+		config.setString(HighAvailabilityOptions.HA_MODE, "ZOOKEEPER");
+		config.setString(CoreOptions.STATE_BACKEND, "ZOOKEEPER");
```
I'm pretty sure, this should have been `FILESYSTEM` and I was just to eager to change everything to `ZOOKEEPER` when copying the tests over... Anyway, this option is not used at all because we only test the BLOB service classes (also in the test classes I copied this from).

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java:
```diff
@@ -415,13 +393,17 @@ private BlobKey putBuffer(@Nullable JobID jobId, byte[] value, int offset, int l
 	 * 		thrown if an I/O error occurs while uploading the data to the BLOB server
 	 */
-	private BlobKey putInputStream(@Nullable JobID jobId, InputStream inputStream) throws IOException {
+	BlobKey putInputStream(@Nullable JobID jobId, InputStream inputStream, boolean permanentBlob)
```
I'm not sure, it's worth it since this method is about to become private to `TransientBlobCache` once the BLOB upload is not necessary from the client anymore, i.e. once we upload the job in a single (HTTP) operation

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobClient.java:
```diff
@@ -601,7 +564,39 @@ public void deleteInternal(@Nullable JobID jobId, BlobKey key) throws IOExceptio
+			if (cause == null) {
+				return false;
+			} else {
+				throw new IOException("Server side error: " + cause.getMessage(), cause);
```
I'm tempted to leave it that way not only because previous code followed that pattern (as in `BlobClient#receiveAndCheckGetResponse()`) but also because it is nice to have a brief failure message with some details without having to dig through the stack trace.

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -389,95 +413,332 @@ public File getFile(JobID jobId, BlobKey key) throws IOException {
+				blobStore.get(jobId, blobKey, incomingFile);
+
+				BlobUtils.moveTempFileToStore(
+					incomingFile, jobId, blobKey, localFile, readWriteLock.writeLock(), LOG, null);
```
Agreed - it's not too nice. The cost of changing that would be another try-catch block duplication at every use of this method though, so all usages would have this instead:
```
writeLock.lock();
try {
	BlobUtils.moveTempFileToStore(incomingFile, jobId, blobKey, storageFile, LOG, blobStore);
} finally {
	writeLock.unlock();
}
```

`BlobUtils` on the other hand is already quiet tightly coupled with the `BlobServer`/`Cache` classes anyway with mostly package-private methods for only those ... not quite sure what would be better and went this way.

tillrohrmann replied:
Alright.

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -0,0 +1,429 @@
+	 *
+	 * @see #registerJob(JobID)
+	 */
+	public void releaseJob(JobID jobId) {
```
With the current code, it wouldn't have mattered, whether the counter is below zero since the cleanup logic inside `PermanentBlobCache#run()` accounted for that and `releaseJob` only set the cleanup time once. It should, however, also considered an improper use case and we may guard against it.

tillrohrmann replied:
I was just wondering, because what happens if it is negative and then this job is registered again? Will the counter then be at `<= 0`? This would mean that all blobs associated with this job will be deleted at the next clean up cycle, right?

NicoK replied:
This behaviour changed with the last commits - the reference counter can now not go below 0 and issues a warning in case it would.

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/TransientBlobCache.java:
```diff
@@ -0,0 +1,322 @@
+	 */
+	public TransientBlobCache(
+			final InetSocketAddress serverAddress,
+			final Configuration blobClientConfig) throws IOException {
```
Yes, that's mostly why. And passing all parameters that the `BlobClient` requires is also not too nice, cumbersome and error-prone.

NicoK commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/TransientBlobCache.java:
```diff
@@ -0,0 +1,322 @@
+		File incomingFile = createTemporaryFilename();
+		try {
+			BlobClient.downloadFromBlobServer(jobId, blobKey, false, incomingFile, serverAddress,
+				blobClientConfig, numFetchRetries);
```
only manually via the `TransientBlobCache#delete()` methods

NicoK commented:
After an offline discussion with @tillrohrmann, we agreed to have more of the transient vs. permanent BLOB handling inside the `BlobServer` by including the type into the `BlobKey`. The caches, however, remain separate (with a common base class for shared code) because of the different cache guarantees they give. In addition, we decided upon having a get-and-delete functionality for transient BLOBs (as planned by a TODO item previously).

This means the following changes:
- the `BlobServer` deletes transient BLOBs after they have been successfully downloaded to a BLOB cache (the client acknowledges that)
- the `TransientBlobCache` and `PermanentBlobCache` differ in the functionality they provide:
  - `PermanentBlobCache` uses ref-counting for cleaning up its resources
  - `TransientBlobCache` imposes the user to manually delete requested BLOB files (from the cache!)
  - `TransientBlobCache` allows uploading (transient) BLOBs
  - `TransientBlobCache` allows non-job-related BLOBs
  - both have a GET method: `#getTransientFile()` vs. `#getPermanentFile()`
  - irrespective of the actual cache being used, the `BlobServer` decides whether to delete the BLOB (transient BLOBs only) or not (permanent BLOBs)
  - theoretically, a permanent BLOB could also be retrieved from the `TransientBlobCache` (and the other way around if job-related) - this behaviour is not used though but basic tests exist to cover it

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobKey.java:
```diff
@@ -83,30 +113,34 @@ public boolean equals(final Object obj) {
 	@Override
 	public int hashCode() {
-		return Arrays.hashCode(this.key);
+		return Arrays.hashCode(this.key) + this.type.hashCode();
```
I think when one combines field, then one should multiply the fields with a prime, e.g. `37`.

NicoK replied:
ok, I wasn't quite sure and based my code on some examples from the current Flink code. Apparently, I took the wrong ones.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -375,7 +378,32 @@ public File getFile(BlobKey key) throws IOException {
 	@Override
-	public File getFile(JobID jobId, BlobKey key) throws IOException {
+	public File getTransientFile(JobID jobId, BlobKey key) throws IOException {
+		checkNotNull(jobId);
```
Should we check that `key` equals actually the `BlobType.TRANSIENT`?

tillrohrmann replied:
What about making the usage of blob keys failsafe by having different types of `BlobKey`. E.g. `TransientBlobKey` and `PermanentBlobKey`?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobServer.java:
```diff
@@ -375,7 +378,32 @@ public File getFile(BlobKey key) throws IOException {
+	 * 		if any other error occurs when retrieving the file
+	 */
+	@Override
+	public File getPermanentFile(JobID jobId, BlobKey key) throws IOException {
```
Same here for `BlobType.PERMANENT`

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobUtils.java:
```diff
@@ -407,4 +419,72 @@ static void closeSilently(Socket socket, Logger LOG) {
+					log.warn("Detected concurrent file modifications. This should only happen if multiple" +
+						"BlobServer use the same storage directory.");
+					// we cannot be sure at this point whether the file has already been uploaded to the blob
+					// store or not. Even if the blobStore might shortly be in an inconsistent state, we have
```
"we have to"

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/FileSystemBlobStore.java:
```diff
@@ -64,39 +65,60 @@ public FileSystemBlobStore(FileSystem fileSystem, String storagePath) throws IOE
+			final byte[] computedKey = md.digest();
+			if (!Arrays.equals(computedKey, blobKey.getHash())) {
+				throw new IOException("Detected data corruption during transfer");
+			}
```
Nice addition :-)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -0,0 +1,290 @@
+	@Override
+	public File getPermanentFile(JobID jobId, BlobKey key) throws IOException {
+		checkNotNull(jobId);
+		return getTransientFileInternal(jobId, key);
```
Maybe we could rename `getTransientFileInternal` into `getFileInternal`.

tillrohrmann commented on flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala:
```diff
@@ -468,7 +468,7 @@ class JobManager(
             case x: ActorTaskManagerGateway =>
               handleTaskManagerTerminated(x.getActorGateway().actor(), instance.getId)
-            case _ => log.debug(s"Cannot remove resource ${resourceID}, because there is " +
+            case _ => log.debug(s"Cannot remove reosurce ${resourceID}, because there is " +
```
typo resource

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheCleanupTest.java:
```diff
@@ -142,7 +128,64 @@ public void testJobCleanup() throws IOException, InterruptedException {
 
 	/**
-	 * Tests that {@link BlobCache} cleans up after calling {@link BlobCache#releaseJob(JobID)}
+	 * Tests that {@link BlobCache} sets the expected reference counts and cleanup timeouts when
```
Rename `BlobCache` to `PermanentBlobCache`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheCleanupTest.java:
```diff
@@ -252,6 +280,36 @@ public void testJobDeferredCleanup() throws IOException, InterruptedException {
+		// because we cannot guarantee that there are not thread races in the build system, we
+		// loop for a certain while until the references disappear
+		{
+			long deadline = System.currentTimeMillis() + 30_000L;
```
Shouldn't the deadline be configurable? `verifyJobCleanup` should not know how long the clean up takes (I guess it depends on the cleanup interval).

NicoK replied:
Since the cleanup interval is only a lower bound on the time of the actual cleanup due to thread scheduling, we can't really use that. Basically, the way, `verifyJobCleanup` is used by now is that one expects that the file is deleted (but can't be sure) and therefore, we have this extra buffer of 30s. It could be made configurable, when desired, I guess, but let's not do this now.

NicoK commented:
Rebased and extended the PR as requested - the last two commits contain the changes compared to the last review.

I tried to clean up some of the commits for a better merge but please note that this PR also includes #4568, #4238 and #4402 fixes and commits. For some of those, changes were applied afterwards in the review process which will cause conflicts in their respective PRs that I will close (fixed after merging this PR).

tillrohrmann commented:
Thanks a lot for your work and patience with me @NicoK. Changes look good to me. I've rebased it onto the latest master and once Travis gives green light, I'll merge it :-)
